### PR TITLE
SLO and ACS Binding are linked with AuthnRequest Binding in SAML Identity Broker Metadata

### LINKED ISSUE #11079: SLO and ACS Binding are linked with AuthnRequest Binding in SAML Identity Broker Metadata
### Describe the bug

Keycloak has three settings whether HTTP-POST Binding is used or not in SAML Identity Broker.
- HTTP-POST Binding Response (postBindingResponse)
- HTTP-POST Binding for AuthnRequest (postBindingAuthnRequest)
- HTTP-POST Binding Logout (postBindingLogout)

But these settings are not reflected in the SP Metadata Binding.

POST Binding Setting

### Version

17.0.1

### Expected behavior

| Keycloak Setting Name |  Keycloak Setting Value  |  SP Metadata Binding  |
| ---- | ---- | ---- |
| postBindingAuthnRequest |  false  |  -  |
| postBindingResponse |  true  |  urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST  |
| postBindingLogout |  true  |  urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST  |

actual behavior

```xml
<md:EntityDescriptor
	xmlns="urn:oasis:names:tc:SAML:2.0:metadata"
	xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata"
	xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
	xmlns:ds="http://www.w3.org/2000/09/xmldsig#" entityID="https://keycloak.example.com/auth/realms/sptest" ID="ID_bffe674a-6bbe-4c8d-9887-61cc7a47ba90">
	<md:SPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol" AuthnRequestsSigned="false" WantAssertionsSigned="false">
		<md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://keycloak.example.com/auth/realms/sptest/broker/samlbinding/endpoint"/>
		<md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:persistent</md:NameIDFormat>
		<md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://keycloak.example.com/auth/realms/sptest/broker/samlbinding/endpoint" isDefault="true" index="1"/>
	</md:SPSSODescriptor>
</md:EntityDescriptor>
```

### Actual behavior

| Keycloak Setting Name |  Keycloak Setting Value  |  SP Metadata Binding  |
| ---- | ---- | ---- |
| postBindingAuthnRequest |  false  |  -  |
| postBindingResponse |  true  |  urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect  |
| postBindingLogout |  true  |  urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect  |

actual behavior

```xml
<md:EntityDescriptor
	xmlns="urn:oasis:names:tc:SAML:2.0:metadata"
	xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata"
	xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
	xmlns:ds="http://www.w3.org/2000/09/xmldsig#" entityID="https://keycloak.example.com/auth/realms/sptest" ID="ID_bffe674a-6bbe-4c8d-9887-61cc7a47ba90">
	<md:SPSSODescriptor protocolSupportEnumeration="urn:oasis:names:tc:SAML:2.0:protocol" AuthnRequestsSigned="false" WantAssertionsSigned="false">
		<md:SingleLogoutService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://keycloak.example.com/auth/realms/sptest/broker/samlbinding/endpoint"/>
		<md:NameIDFormat>urn:oasis:names:tc:SAML:2.0:nameid-format:persistent</md:NameIDFormat>
		<md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://keycloak.example.com/auth/realms/sptest/broker/samlbinding/endpoint" isDefault="true" index="1"/>
	</md:SPSSODescriptor>
</md:EntityDescriptor>
```

### How to Reproduce?

1. Create SAML Identity Provider
2. HTTP-POST Binding Response Setting ON
3. HTTP-POST Binding Logout Setting ON
4. Save
5. Check SAML Identity Provider Metadata

### Anything else?

I found the another problem.
When the setting of postBindingLogout is null, postBindingLogout is linked with the setting of postBindingLogout.

[isPostBindingLogout](https://github.com/keycloak/keycloak/blob/bfce612641a70e106b20b136431f0e4046b5c37f/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java#L222)

```java
        if (postBindingLogout == null) {
            // To maintain unchanged behavior when adding this field, we set the inital value to equal that
            // of the binding for the response:
            return isPostBindingResponse();
```

HTTP-POST Binding Logout default value is OFF in admin console.
Therefore appearance of admin console and actual behavior do not match.
I think that HTTP-POST Binding Logout  default is `ON` in appearance of new admin cosole.

### PR DESCRIPTION
**Pull request by i7a7467**
Closes #11079

### CODE REVIEW

**Comment by mposolda** [comment]
@rmartinc Could you please review this one?
@i7a7467 Are you please able to rebase your PR to latest Keycloak main to fix the conflicts?

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java [hunk 1]
```diff
@@ -343,10 +343,16 @@ private JaxrsSAML2BindingBuilder buildLogoutBinding(KeycloakSession session, Use
     public Response export(UriInfo uriInfo, RealmModel realm, String format) {
         try
         {
-            URI authnBinding = JBossSAMLURIConstants.SAML_HTTP_REDIRECT_BINDING.getUri();
+            URI authnResponseBinding = JBossSAMLURIConstants.SAML_HTTP_REDIRECT_BINDING.getUri();
 
-            if (getConfig().isPostBindingAuthnRequest()) {
-                authnBinding = JBossSAMLURIConstants.SAML_HTTP_POST_BINDING.getUri();
+            if (getConfig().isPostBindingResponse()) {
```
Comment (rmartinc) [inline services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java:348]: I think we should maintain using `isPostBindingAuthnRequest` here. In theory both values should be the same but the one that really is forcing to use the post binding is the `isPostBindingAuthnRequest` (see [here](https://github.com/keycloak/keycloak/blob/21.0.1/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java#L170) and then [here](https://github.com/keycloak/keycloak/blob/21.0.1/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java#L196-L200)). The ` isPostBindingResponse` is only used to set the attribute in the request ([here](https://github.com/keycloak/keycloak/blob/21.0.1/services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java#L135-L137)).

Modify the tests accordingly please.

**Review by rmartinc — CHANGES_REQUESTED** [review]
Thanks for the PR @i7a7467!

It's almost OK, but I think we should maintain `isPostBindingAuthnRequest` in favor of `isPostBindingResponse`.

**Comment by i7a7467** [comment]
@rmartinc 
Sorry for the delay in replying to you. I am checking the link you gave me codes.
`(see here and then here).` Could you please give me a few days?

**Comment by rmartinc** [comment]
@i7a7467 Yes, no problem at all, just wanted to know if you were working on it. Thanks!

**Comment by i7a7467** [comment]
@rmartinc 
Thanks reply.
Sorry for the late reply.
I read comments and code you provided.

The setting of `HTTP-POST binding for AuthnRequest` does not affect SAML Service Provider metadata.
There I don't know the background behind your idea of changing this code.
Could you give me some more details?

My opinion is as follows.

### HTTP-POST binding response
- Off
  - The value of `AssertionConsumerService Binding` in metadata
    - `urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect`
  - The value of `ProtocolBinding` in authentication request assertion to IdP
    - `urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect`
- On
  - The value of `AssertionConsumerService Binding` in metadata
    - `urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST`
  - The value of `ProtocolBinding` in authentication request assertion to IdP
    - `urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST`

metadata

authentication request assertion

### HTTP-POST binding for AuthnRequest
- Off
  - Use GET HTTP method in authentication request to IdP
- On
  - Use POST HTTP method in authentication request to IdP

image

**Comment by rmartinc** [comment]
@i7a7467 There are three different options you are managing here:

* HTTP-POST binding response (isPostBindingResponse): This one is just used to indicate the binding type in the XML payload.
* HTTP-POST binding for AuthnRequest (isPostBindingAuthnRequest): This one is used to really send a POST or GET/REDIRECT request for the authn. If true a POST is used, if false REDIRECT is used. Check the links I presented before.
* HTTP-POST binding logout (isPostBindingLogout): This one is clear, true use POST for logout, false use REDIRECT.

As you see no option is directly involved with the descriptor. They are used to choose what BINDING to use when doing the different requests. The option that is really used to send the POST binding for the authn request  is the second one (isPostBindingAuthnRequest). The first one (isPostBindingResponse) is only used to add the XML binding in the XML payload. If you configure one with true and the other with false, you can be sending a POST (using POST binding) but assigning to the XML payload that it is a REDIRECT binding (so a nonsense). IMHO both options ("HTTP-POST binding response" and "HTTP-POST binding for AuthnRequest") should be always the same and one can even be removed. I would always use same value in the tests for those two options.

But, as you say, the three options are not directly involved with the descriptor. My request is to maintain `isPostBindingAuthnRequest` as it was done before just for backwards compatibility. I'm OK with changing the logout as you are doing but I prefer to maintain `isPostBindingAuthnRequest` as before.

**Comment by i7a7467** [comment]
@rmartinc 
Thanks for the details.
I understand about maintaining backward compatibility. 
And I also understood about wanting to keep the behavior the same as it is today.
I will modify codes for backward compatibility.

> IMHO both options ("HTTP-POST binding response" and "HTTP-POST binding for AuthnRequest") should be always the same and one can even be removed. I would always use same value in the tests for those two options.

However, I do not understand the opinion that `isPostBindingResponse` and `isPostBindingAuthnRequest` should also have the same settings.
HTTP-Redirect or HTTP-POST is often used for authenrequest , while HTTP-POST is generally used for authenresponse.(e.g., Google Workspace and box).
I think it is necessary to allow separate Binding for authenrequest  and autheresponse.

**Review by rmartinc — APPROVED** [review]
Thanks for the change @i7a7467 ! LGTM.

**Comment by mposolda** [comment]
@i7a7467 @rmartinc Thanks for the work and review!

**Review by mposolda — APPROVED** [review]

**Pull request merged**